fix(client): Serialize all Floats in exponential notation

cmd-johnson opened the pull request:
This should fix #12651, #13317 and #15551 by always serializing Float fields in GraphQL queries in exponential notation.

I added a unit test for the serialization and a integration test that verifies that large floats can now be inserted into the database and updated other tests that used floats to match the new serialization behavior.

In order to properly test this, I also had to change the regex used by the `jestSnapshotSerializer` in `normalizeMigrateTimestamps`, because it interfered with all long numbers in snapshots. E.g. `-9.007199254740991e+15` would get replaced with `-9.202012310000001e+15`, which made snapshots appear to test for a wrong value. I updated snapshots that had the same problem accordingly.

Fixes #12651
Fixes #13317
Fixes #15551

cmd-johnson committed:
test(client): unit test for #12651

cmd-johnson committed:
test(client): integration test for #12651

cmd-johnson committed:
fix(client): Serialize all floats in exponential notation

Fix #12651, fix #13317

cmd-johnson committed:
test(internals): Fix normalizeMigrateTimestamps applying to all long numbers

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma?pullRequest=15879) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma?pullRequest=15879) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma?pullRequest=15879) it.</sub>

SevInf committed:
test(client): Convert large-floats tests to functional

SevInf committed:
test(client): Revert snapshot serializer changes

It does not seem that it was indendened for snapshot serizier to apply
to select.test.ts. Instead, we are fixing the failure by changing the
number so it won't be replaced by serializer.

SevInf reviewed (approved):
Hi @cmd-johnson!
Thank you for the great contribution, especially adding both unit & integration tests!

I've made a couple of changes to the PR:
- converted intergration test to a [functional](https://github.com/prisma/prisma/blob/main/TESTING.md#functional-tests-for-the-client) one. This a newer, more powerful test suite and we generally add new tests there and consider integration suite to be a legacy.

Once CI is green, I'll merge your contribution. Then, it will be released as a part of prisma 4.6.0.

SevInf committed:
Merge branch 'main' into fix/float-serialization

SevInf committed:
Revert "test(client): Revert snapshot serializer changes"

This reverts commit 7756be2edd37d48ab8c00e612c9e9190092184f1.

aqrln commented:
Also fixes https://github.com/prisma/prisma/issues/15551

aqrln commented on packages/internals/src/utils/jestSnapshotSerializer.js:
```diff
@@ -69,7 +69,8 @@ function normalizeBinaryFilePath(str) {
 function normalizeMigrateTimestamps(str) {
-  return str.replace(/\d{14}/g, '20201231000000')
+  return str.replace(/(?<!\d)\d{14}(?!\d)/g, '20201231000000')
+  // return str.replace(/\d{14}/g, '20201231000000')
```
```suggestion
```

SevInf committed:
Update packages/internals/src/utils/jestSnapshotSerializer.js

Co-authored-by: Alexey Orlenko <alex@aqrln.net>
